pkg/backoff/backoff.go

```diff
@@ -0,0 +1,123 @@
+package backoff
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	ini "github.com/wakatime/wakatime-cli/pkg/config"
+	"github.com/wakatime/wakatime-cli/pkg/heartbeat"
+	"github.com/wakatime/wakatime-cli/pkg/log"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	// resetAfter sets the total seconds a backoff will last.
+	resetAfter = 3600
+	// factor is the total seconds to be multiplied by.
+	factor = 15
+)
+
+// RetryConfig defines backoff data.
+type RetryConfig struct {
+	// At is the time when the first failure happened.
+	At *time.Time
+	// Retries is the number of attempts to connect.
+	Retries int
+	// V is an instance of Viper.
+	V *viper.Viper
+}
+
+// WithRetry initializes and returns a heartbeat handle option, which
+// can be used in a heartbeat processing pipeline to prevent trying to send
+// a heartbeat when the api is unresponsive.
+func WithRetry(config RetryConfig) heartbeat.HandleOption {
+	return func(next heartbeat.Handle) heartbeat.Handle {
+		return func(hh []heartbeat.Heartbeat) ([]heartbeat.Result, error) {
+			log.Debugln("execute heartbeat backoff algorithm")
+
+			nextDuration, _ := time.ParseDuration(fmt.Sprintf("%fs", float64(factor)*math.Pow(2, float64(config.Retries))))
+
+			now := time.Now()
+			if config.At != nil && now.Before(config.At.Add(nextDuration)) {
```

Comment:
I guess we want to include `resetAfter` here, to ensure, that we send the heartbeats after the defined time. Otherwise we will not reset after 3600 seconds.
```suggestion
			if config.At != nil && now.Before(config.At.Add(nextDuration)) && now.Before(config.At.Add(resetAfter)) {
```

Reply:
I believe we don't need to ensure that current time is before last backoff time because at `updateBackoffSettings` it resets backoff in case of current time is greater than the reset limit.

```Go
// reset backoff settings if reached max allowed time
if at != nil && now.Sub(*at).Seconds() > resetAfter {
	keyValue["backoff_at"] = now.Format(ini.DateFormat)
	keyValue["backoff_retries"] = "1"
}
```

Reply:
Shouldn't this reset to zero instead of `1`?

Reply:
Also, we should set `backoff_at ` to null when resetting to have no backoff on the next try.

Reply:
We never save `retries = 0`, so cleaning retries will make cli to consider it as zero value. @alanhamlett reset will be fixed to not save current time. Thanks for pointing out.

Reply:
Latest changes fix the problem, which I pointed out :+1: